packages/core/core-flows/src/product/workflows/update-products.ts

```diff
@@ -1,26 +1,167 @@
+    sales_channels?: { id: string }[]
+  })[]
+}
+
 type UpdateProductsStepInput =
-  | {
-      selector: ProductTypes.FilterableProductProps
-      update: ProductTypes.UpdateProductDTO
-    }
-  | {
-      products: ProductTypes.UpsertProductDTO[]
-    }
+  | UpdateProductsStepInputSelector
+  | UpdateProductsStepInputProducts
 
 type WorkflowInput = UpdateProductsStepInput
 
+function prepareUpdateProductInput({
+  input,
+}: {
+  input: WorkflowInput
+}): UpdateProductsStepInput {
+  if ("products" in input) {
+    return {
+      products: input.products.map((p) => ({
+        ...p,
+        sales_channels: undefined,
+      })),
+    }
+  }
+
+  return {
+    selector: input.selector,
+    update: {
+      ...input.update,
+      sales_channels: undefined,
+    },
+  }
+}
+
+function updateProductIds({
+  updatedProducts,
+  input,
+}: {
+  updatedProducts: ProductTypes.ProductDTO[]
+  input: WorkflowInput
+}) {
+  if ("products" in input) {
+    let productIds = updatedProducts.map((p) => p.id)
+    const discardedProductIds: string[] = input.products
+      .filter((p) => !p.sales_channels)
+      .map((p) => p.id as string)
+    return arrayDifference(productIds, discardedProductIds)
+  }
+
+  return !input.update.sales_channels ? [] : undefined
+}
+
+function prepareSalesChannelLinks({
+  input,
+  updatedProducts,
+}: {
+  updatedProducts: ProductTypes.ProductDTO[]
+  input: WorkflowInput
+}): Record<string, Record<string, any>>[] {
+  if ("products" in input) {
+    return input.products
+      .filter((p) => p.sales_channels)
+      .flatMap((p) =>
+        p.sales_channels!.map((sc) => ({
+          [Modules.PRODUCT]: {
+            product_id: p.id,
+          },
+          [Modules.SALES_CHANNEL]: {
+            sales_channel_id: sc.id,
+          },
+        }))
+      )
+  }
+
+  if (input.selector && input.update.sales_channels?.length) {
+    return updatedProducts.flatMap((p) =>
+      input.update.sales_channels!.map((channel) => ({
+        [Modules.PRODUCT]: {
+          product_id: p.id,
+        },
+        [Modules.SALES_CHANNEL]: {
+          sales_channel_id: channel.id,
+        },
+      }))
+    )
+  }
+
+  return []
+}
+
+function prepareToDeleteLinks({
+  currentLinks,
+}: {
+  currentLinks: {
+    product_id: string
+    sales_channel_id: string
+  }[]
+}) {
+  return currentLinks.map(({ product_id, sales_channel_id }) => ({
+    [Modules.PRODUCT]: {
+      product_id,
+    },
+    [Modules.SALES_CHANNEL]: {
+      sales_channel_id,
+    },
+  }))
+}
+
 export const updateProductsWorkflowId = "update-products"
 export const updateProductsWorkflow = createWorkflow(
   updateProductsWorkflowId,
   (
     input: WorkflowData<WorkflowInput>
   ): WorkflowData<ProductTypes.ProductDTO[]> => {
     // TODO: Delete price sets for removed variants
-    // TODO Update sales channel links
-    return updateProductsStep(input)
+
+    const toUpdateInput = transform({ input }, prepareUpdateProductInput)
```

Comment:
thought: I don't know if I like, that the steps that were added here are so specific to sales channel management. E.g. as soon as we add another link to products, we will need to revisit all of these. Or create separate steps.

Reply:
I need to relook into it but your comment is on the transformer, are you referring to the transformer of the next step or the step itself just to be sure